scripts/field_report.py

```diff
@@ -0,0 +1,262 @@
+    def __init__(self, msg: str) -> None:
+        super().__init__('Sanity check failed: ' + msg +
+                         '\nThis likely indicates a compilation error.' +
+                         '\nMaybe a missing include directory (-I)?')
+
+
+
+class FieldInfo:
+    """Information about a field of a structure."""
+
+    def __init__(self, node: Cursor) -> None:
+        self.node = node
+        self.lexical_uses = 0
+        self.size = node.type.get_size()
+        self.align = node.type.get_align()
+        # Empirically, offsetof is in bits, not bytes. To make the output
+        # easier to read, convert to bytes (the same unit as size and
+        # alignment), which means that bitfields will be located at their
+        # first byte.
+        self.offset = node.get_field_offsetof() // 8
+
+    def record_use(self, _node: Cursor) -> None:
+        self.lexical_uses += 1
+
+    def uses(self) -> int:
+        return self.lexical_uses
+
+    def is_indirect(self) -> bool:
+        """Whether access to this field is indirect on ARM Cortex-M0+."""
+        word_offset = self.offset // self.align
+        return word_offset >= 128
+
+    def score(self) -> int:
+        """A field's score is an estimate of its access cost.
+
+        The cost is calculated as the number of instructions needed to
+        access the field on ARM Cortex-M0+, multiplied by the number of uses
+        of the field inside the library.
+        """
+        cost_per_use = 1 + 2 * self.is_indirect()
+        return self.uses() * cost_per_use
+
+
+class Ast:
+    """Abstract representation of the source code."""
+
+    def __init__(self, options) -> None:
+        """Prepare for analysis of C source files."""
+        if options.clang_library_file:
+            clang.cindex.Config.set_library_file(options.clang_library_file)
+        self.parse_options = []
+        if options.target:
+            self.parse_options += ['-target', options.target]
+        for d in options.include:
+            self.parse_options.append('-I' + d)
+        for d in options.define:
+            self.parse_options.append('-D' + d)
+        self.index = clang.cindex.Index.create()
+        self.files = {} #type: Dict[str, TranslationUnit]
+        # fields[TYPE_OR_STRUCT_NAME][FIELD_NAME]
+        self.fields = {} #type: Dict[str, Dict[str, FieldInfo]]
+
+    def load(self, *filenames: str) -> None:
+        """Load the AST of the given source files."""
+        for filename in filenames:
+            self.files[filename] = self.index.parse(filename,
+                                                    self.parse_options)
+
+    def in_interesting_file(self, location: SourceLocation) -> bool:
+        """Whether the given location is in a file that should be analyzed."""
+        if not hasattr(location.file, 'name'):
+            # Some artificial nodes have associated no file name.
+            # Let's hope they're not important.
+            return False
```

Comment:
Given that we are not sure about the files that miss this attribute, it could be worth it to capture the list somewhere. Or add an optional verbosity level on the parser that would print them to the terminal.

Reply:
When a node has no location, it's difficult to analyze what it's for. I don't think we could print useful information in this log.